python_modules/dagster/dagster_tests/execution_tests/engine_tests/multienvironment_tests/asset_variant/test_multi_environment_asset_execution.py

```diff
@@ -0,0 +1,143 @@
+from dagster._core.definitions.op_definition import OpDefinition
+from dagster._core.definitions.reconstruct import reconstructable
+from dagster._core.definitions.unresolved_asset_job_definition import define_asset_job
+from dagster._core.execution.api import execute_job
+from dagster._core.execution.retries import RetryMode
+from dagster._core.executor.multi_environment.multi_environment_step_handler import (
+    MultiEnvironmentExecutor,
+    RemoteEnvironmentSingleStepHandler,
+)
+from dagster._core.test_utils import instance_for_test
+from dagster._utils import file_relative_path
+
+from dagster_tests.execution_tests.engine_tests.multienvironment_tests.python_script_single_step_handler import (
+    PythonScriptSingleStepHandler,
+)
+
+
+class RemoteAsset(NamedTuple):
+    asset_key: CoercibleToAssetKey
+    upstream_asset_keys: Sequence[CoercibleToAssetKey]
+    single_step_handler: RemoteEnvironmentSingleStepHandler
+
+
+def remote_asset_in_script(asset_key: str, upstream_asset_keys: Sequence[str]):
+    return RemoteAsset(
+        asset_key=asset_key,
+        upstream_asset_keys=upstream_asset_keys,
+        # you can customize step handling per step
+        single_step_handler=PythonScriptSingleStepHandler(
+            file_relative_path(__file__, f"{asset_key}.py")
+        ),
+    )
+
+
+def build_definitions_for_remote_assets(
+    asset_job_name: str, remote_assets: Sequence[RemoteAsset]
+) -> Definitions:
+    def _never_called(*args, **kwargs):
+        raise Exception("This assets exists only for metadata. Do not call")
+
+    assets_defs = []
+
+    for remote_asset in remote_assets:
+        asset_key = AssetKey.from_coerceable(remote_asset.asset_key)
+        upstream_asset_keys = [
+            AssetKey.from_coerceable(uak) for uak in remote_asset.upstream_asset_keys
+        ]
+        keys_by_input_name = {
+            upstream_asset_key.to_python_identifier(): upstream_asset_key
+            for upstream_asset_key in upstream_asset_keys
+        }
+        keys_by_output_name = {"result": asset_key}
+        op_def = OpDefinition(
+            compute_fn=_never_called,
+            name=asset_key.to_user_string(),
+            ins={
+                upstream_asset_key.to_python_identifier(): In()
+                for upstream_asset_key in upstream_asset_keys
+            },
+        )
+        assets_defs.append(
+            AssetsDefinition(
+                keys_by_input_name=keys_by_input_name,
+                keys_by_output_name=keys_by_output_name,
+                node_def=op_def,
+            )
+        )
+
+    assets_to_handlers = {
+        remote_asset.asset_key: remote_asset.single_step_handler for remote_asset in remote_assets
+    }
+
+    return Definitions(
+        assets=assets_defs,
+        jobs=[define_asset_job(asset_job_name, "*")],
+        executor=MultiEnvironmentExecutor(
+            lambda step_context: assets_to_handlers[step_context.step.key],
```

Comment:
and here it is the responsibility of the user to write a function that maps a step context in the launching process to a remove step handler

Reply:
Is this something we would consider supporting via config or tags? it seems pretty powerful to be able to say something like "by default every step for this job should run in a separate k8s pod, except for this step which should run in a subprocess" without needing to write a lambda

And doing the mapping in user code locks us in to run workers being a user code process right? Which is probably fine at this point, but just flagging..

Reply:
I think it would be user code in the same way that DagsterInstance is "user code" because of the ConfigurableClass stuff.

Reply:
Re: supporting tags I think that is a great idea.